### ISSUE TITLE
Nullable analysis of default parameter values

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#32844]
We currently do a simple, but incomplete, analysis of default parameter values.

```C#
        [WorkItem(26626, "https://github.com/dotnet/roslyn/issues/26626")]
        [Fact]
        public void ParameterDefaultValue_02()
        {
            var source =
@"class C
{
    const string? S0 = null;
    static void F0(string s = null!) { }
    static void F1(string x = (string)null!, string y = ((string)null)!) { }
    static void F2(string x = default!, string y = default(string)!) { }
    static void F3(string s = (S0!)!) { }
    static void M()
    {
        F0();
        F1();
        F2();
        F3();
        F1(x: null);
        F1(y: null);
        F2(null!, null);
    }
}";
            // TODO2 we should warn on the conversion of `null` to `string` in `(string)null`
            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
            comp.VerifyDiagnostics(
                // (14,15): warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.
                //         F1(x: null);
                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(14, 15),
                // (15,15): warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.
                //         F1(y: null);
                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(15, 15),
                // (16,19): warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.
                //         F2(null!, null);
                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(16, 19)
                );
        }
```

**Comment by RikkiGibson** [comment]
We are closer in 16.8 but I think this won't be quite right until parameter default values are part of the bound tree (we can visit the bound default value simply as an expression, and then check assignment of the "lowered" constant value to the parameter).

**Comment by RikkiGibson** [comment]
Closing as only very minor aspects of this issue are still unsolved. /cc @jcouv
